### PR TITLE
Self protocore setup.

### LINKED ISSUE #189: Implement SelfProtocore::setup()
1. Create the `SelfProtocore` contract

2. Implement `SelfProtocore::setup()` 
    - This will be an external function
    - This function initializes the genesis link from values defined in `GenesisSelfProtocore`
        - `sourceTransitionHash` will be set to GENESIS_SOURCE_TRANSITION_HASH
        - `parentVoteMessageHash` will be set to GENESIS_PARENT_VOTE_MESSAGE_HASH
        - `targetBlockHash` from GENESIS_TARGET_BLOCK_HASH
        - `targetBlockNumber` from GENESIS_TARGET_BLOCK_NUMBER
    - `targetFinalisation` of the genesis link will be set to finalised.
    - `proposedMetablockHeight` will be set to 0.
    - `forwardVoteCount` -s will be set to 0.
    - The function can be called only by coconsensus contract.
    - The function can be called only once.

### PR DESCRIPTION
**Pull request by deepesh-kn**
This PR is blocked by : #227 
This pull request contains the following changes:
1. Implementation of `SelfProtocore` contract.
2. Updated storage variables in `GenesisSelfProtocore` contract.
3. Added unit tests for `SelfProtocore::setup`

fixes: #189

### CODE REVIEW

File: contracts/protocore/SelfProtocore.sol [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        bytes32 _metachainId,
+        bytes32 _domainSeparator,
+        uint256 _epochLength,
+        uint256 _metablockHeight
+    )
+        external
+        onlyCoconsensus
+    {
+        require(
+            epochLength == uint256(0),
```
Comment (pgev) [inline contracts/protocore/SelfProtocore.sol]: I would rather use _domainSeparator, however, this one works also :).

File: contracts/protocore/Protocore.sol [hunk 2]
```diff
@@ -17,14 +17,13 @@ pragma solidity >=0.5.0 <0.6.0;
...
 import "../consensus/CoconsensusModule.sol";
-import "../validator/ValidatorSet.sol";
 import "../version/MosaicVersion.sol";
 
 /**
  * @title Protocore abstract contract acting as a base contract for
  *        OriginProtocore and SelfProtocore contracts.
  */
-contract Protocore is MosaicVersion, ValidatorSet, CoconsensusModule {
+contract Protocore is MosaicVersion, CoconsensusModule {
```
Comment (pgev) [inline contracts/protocore/Protocore.sol:26]: Why it's not ValidatorSet anymore?

**Review by pgev — COMMENTED** [review]
@deepesh-kn thx for changes. As this PR has dependency on this one https://github.com/mosaicdao/mosaic-1/pull/227, it's difficult to review and separate what change is included in which one. I've added some comments (I am not going to either approve or request changes) you can check. Once the other PR is merged we need to re-review this one.

File: contracts/protocore/Protocore.sol (hunk 2 above)
Comment (deepesh-kn) [reply contracts/protocore/Protocore.sol:26]: I think that the `ValidatorSet` should be in `SelfProtocore` and not `Protocore`.
`SelfProtocore` and `OriginProtocore` inherits the `Protocore` contract.
As per the specifications, `OriginProtocore` will not have validator sets and will rely on the `SelfProtocore` for any validator related queries. So the `SelfProtocore` will inherit `ValidatorSet` and not `Protocore`. We can discuss this point, else I can revert the changes.

File: contracts/protocore/SelfProtocore.sol (hunk 1 above)
Comment (deepesh-kn) [reply contracts/protocore/SelfProtocore.sol]: I removed this check.
Added a check in `Protocore::setup`
```
require(
            metachainId == bytes32(0),
            "Contract is already initialized."
        );
```

File: contracts/protocore/Protocore.sol (hunk 2 above)
Comment (pgev) [reply contracts/protocore/Protocore.sol:26]: I do see your point. However, Casper game is played at Protocore (abstract). There we need to know does validator belongs to forward validator set of the given metablock height, or, what is the forward validators count at the given metablock height. All this are provided by ValidatorSet implementation. We can handle this probably in the https://github.com/mosaicdao/mosaic-1/issues/182 ticket by introducing a ValidatorSet interface.

**Review by pgev — APPROVED** [review]
Looks good to me!

**Pull request merged**